rumqttd/src/router/mod.rs

```diff
@@ -86,6 +86,34 @@ pub enum Notification {
     Unschedule,
 }
 
+type MaybePacket = Option<Packet>;
+
+// We either get a Packet to write to buffer or we unschedule which is represented as `None`
+impl From<Notification> for MaybePacket {
+    fn from(notification: Notification) -> Self {
+        let packet: Packet;
+        match notification {
+            Notification::Forward(forward) => {
+                packet = Packet::Publish(forward.publish, None);
+            }
+            Notification::DeviceAck(ack) => match ack {
+                Ack::ConnAck(_, connack) => packet = Packet::ConnAck(connack, None),
+                Ack::PubAck(puback) => packet = Packet::PubAck(puback, None),
+                Ack::SubAck(suback) => packet = Packet::SubAck(suback, None),
+                Ack::PingResp(pingresp) => packet = Packet::PingResp(pingresp),
+                Ack::PubRec(pubrec) => packet = Packet::PubRec(pubrec, None),
+                Ack::PubRel(pubrel) => packet = Packet::PubRel(pubrel, None),
+                Ack::PubComp(pubcomp) => packet = Packet::PubComp(pubcomp, None),
+                Ack::UnsubAck(unsuback) => packet = Packet::UnsubAck(unsuback, None),
+                _ => unimplemented!(),
```

Comment:
return a `ParseError` of some kind? Or as return type ( `MaybePacket` ) is an `Option`, just return `None`!

Reply:
None holds special significance here (it implies that the the Link reading this MaybePacket should read again from the Router - so wouldn't want to use it). Restructuring this would take some more thought, so running with a `unimplemented()` for now